tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8899,6 +8917,18 @@ static bool ConvertComponent(ArTypeInfo TargetInfo, ArTypeInfo SourceInfo,
           ComponentConversion = ICK_Floating_Integral;
       }
     }
+  } else if (TargetInfo.EltTy != SourceInfo.EltTy) {
+    // Types are identical in HLSL, but not identical in clang,
+    // such as unsigned long vs. unsigned int.
+    // Add conversion based on the type.
+    if (IS_BASIC_AINT(TargetInfo.EltKind))
```

Comment:
Is this not something the integer promotion rules should handle? I don't remember which way it goes, but in C `int` either promotes to `long` or vice versa. I think we should be able to use the C promotion rules here to get the result we need.

Reply:
My understanding is that we are overriding the type promotion with our own, so if we don't insert the cast here, nobody will, and codegen will assert due to a technical mismatch in types, even though they are equivalent.